Improve Python health check

blueyed opened the pull request:
- handle version from `pip install --user -e`
- report error from the `import` failure

Side notes:
- there should probably be a `neovim.VERSION` that should be used instead of the manual parsing!
- we could report the location of the found module, e.g. via `inspect.getfile()`

TODO:
 - try to import and use `neovim.VERSION`

blueyed committed:
s:check_python: handle 'pip install --user -e'

blueyed committed:
Improve error reporting for Python health check

 - s:version_info: return errors from the `import`
 - skip report_ok for latest version, in case the current version was
   not found.  status contains the error that has been reported in that
   case already.

justinmk commented:
> there should probably be a neovim.VERSION that should be used instead of the manual parsing!

You mean for the client version? Or Nvim version? Will https://github.com/neovim/neovim/pull/5386 help?

tweekmonster commented:
> Will #5386 help?

This is just for describing Neovim's API itself right?  If so, I think it'd be more helpful if the host client actually contained *its* version within the source (e.g. `neovim.__version__` for the python client).

blueyed commented:
Yes, I suggest using/adding `neovim.VERSION` for this.

bfredl commented:
https://github.com/neovim/python-client/pull/232 has a proposal for this. It has both `neovim.VERSION` for the version of the python module and `nvim.version` for the version of the connected nvim instance (for 0.1.6 and later).

blueyed commented:
Should I do something here, apart from rebasing?

blueyed commented:
@mhinz via IRC:
> Essentially you only add `.egg-info/PKG-INFO`, so it seems fine to me.

justinmk commented:
Needs a rebase.

blueyed commented:
Rebased, hopefully not messed up.

blueyed committed:
healthcheck: python: include nvim path for unknown/outdated version

This helps to identify where the `neovim` module is coming from, e.g.

    - INFO: python3-neovim version: 0.1.10 (outdated; from ~/Vcs/neovim-python-client/neovim)
    - WARNING: Latest python3-neovim is NOT installed: 0.1.12

blueyed commented:
Rebased, and added information about the `neovim` Python module's path in case of outdated/unknown versions.

mhinz commented:
LGTM.

Do you mind adding another commit that removes a bit verbosity? E.g. in my case it looks like this:

```
  - INFO: Using: g:python3_host_prog = "/usr/local/bin/python3"
  - INFO: Executable: /usr/local/bin/python3
  - INFO: Python3 version: 3.6.0
  - INFO: python3-neovim version: 0.1.12
  - SUCCESS: Latest python3-neovim is installed: 0.1.12
```

Lines 2 and 4 are unneeded here.

blueyed committed:
healthcheck: python: prefer neovim.VERSION

blueyed committed:
healthcheck: s:check_python: only report latest with good response

blueyed committed:
healthcheck: s:download: do not report curl errors (twice)

Since `curl -s` is used, the actual error (e.g. "curl: (6) Could not
resolve host: pypi.python.org") gets not reported anyway, and s:download
returns an error message with `s:shell_error` already.
This changes `s:download` to also include `a:url` in the error message
now.

It removes

> ERROR: Command error (7) curl -sL https://pypi.python.org/pypi/neovim/json:

And changes

> ERROR: HTTP request failed: error: curl error with https://pypi.python.org/pypi/neovim/json: 6

blueyed commented:
@mhinz 

    - INFO: Executable: /usr/local/bin/python3
    - INFO: Python3 version: 3.6.0

should be

    - INFO: Python3 version: 3.6.0 (from /usr/local/bin/python3)

Agreed that line 4 should be removed.

Also agreed that this should be another PR.

blueyed commented:
Can somebody restart https://travis-ci.org/neovim/neovim/jobs/190045776, in case it's a temporary failure?

justinmk commented on runtime/autoload/health/provider.vim:
```diff
@@ -375,29 +383,33 @@ function! s:check_python(version) abort
-      call health#report_ok(printf('Latest %s-neovim is installed: %s',
-            \ python_bin_name, latest))
+    else
+      if s:is_bad_response(status)
```
can avoid the extra nesting level, just make this `elseif`.

justinmk commented:
@blueyed LGTM, needs one last rebase (sorry, I just pushed a Windows fix to health.vim which conflicts with this)

blueyed commented:
Not sure about the conflict, I've kept HEAD:
```diff
<<<<<<< HEAD
||||||| parent of 041fe250... Improve error reporting for Python health check
      let suggestions = [
            \ 'Error found was: ' . current,
            \ 'Use the command `$ pip' . a:version . ' install neovim`',
            \ ]
=======
      let suggestions = [
            \ 'The error was: ' . status,
            \ 'To install the Python client use: `pip' . a:version . ' install neovim`',
            \ ]
>>>>>>> 041fe250... Improve error reporting for Python health check
```

justinmk commented:
Great improvement, thanks @blueyed !
